### ISSUE TITLE
tfm corrupts files being a multiple of 4096 bytes + 1 and ending with "0"

### ISSUE THREAD

**Issue by divinity76** [prasathmani/tinyfilemanager#1016]
the while loop in https://github.com/prasathmani/tinyfilemanager/blob/6a6eb8abecef523aafda0e9aa71f94f231128222/tinyfilemanager.php#L972 

has a bug where, if a file has filesize a multiple of 4096 bytes + 1, and the file ends with `0` , then the loop will exit early, and the last `0` will be omitted from the copy: https://3v4l.org/PItRM

proposed fix in https://github.com/prasathmani/tinyfilemanager/pull/1014
```php
while (!in_array($buff = fread($in, 4096), array("", false), true)) { fwrite($out, $buff); }
```
alternatively, if you're into micro-optimizations, doing
```php
for(;;){
    $buff = fread($in, 4096);
    if($buff === false || $buff === ""){
        break;
    }
    fwrite($out, $buff);
}
```
would use fewer cpu cycles, avoiding a relatively expensive function call (in_array).. but meh, bet you'd have to use [hyperfine](https://github.com/sharkdp/hyperfine) to even notice a difference.

**Comment by ner00** [comment]
Could you give an example of such filesize? I can't seem to find any multiple of 4096 +1 that ends in 0.

**Comment by ner00** [comment]
I misunderstood, I assumed the last byte was literally hex `0x00` instead of the character `0` (`0x30`), but even though the test was failing for me the logic of the bug you explained made sense logically. I was just trying to understand why I wasn't being able to replicate it, now I know.

As for the usage of `in_array`, do you think something like below would be any better or not really?
```php
while (!feof($in)) { $buff = fread($in, 4096); fwrite($out, $buff); }
```

**Comment by divinity76** [comment]
@ner00 i think that loop is even better :) the usual case where fread() returns emptystring before feof() is with [non-blocking streams / stream-set-blocking()](https://www.php.net/manual/en/function.stream-set-blocking), but we are not using non-blocking IO, we are using blocking IO, but even then i think there are some networked filesystems where fread() can return emptystring before feof, and your loop might increase compatibility with some weird filesystems.

(One example is Microsoft Windows's [WebDAV](https://en.wikipedia.org/wiki/WebDAV) mounted file systems, if you mount a WebDAV on Microsoft Windows, fread() can return emptystring before eof even with blocking IO... it makes many programs malfunction, for example LibreOffice Calc )

**Comment by ner00** [comment]
This one should be closed seeing as fix was implemented.

**Comment by divinity76** [comment]
fixed in f6a93653bee0afe9b437cf5d59f950a2fe2b98cc, closing. (and optimized in f7a2f770081faf5cbfbf7835d6a803470c3c6d5a )
@ner00 thanks for the reminder ^^
